core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -362,11 +382,37 @@ public void finalizeRecovery() {
         cancellableThreads.execute(recoveryTarget::finalizeRecovery);
 
         if (isPrimaryRelocation()) {
+            // in case of primary relocation we have to ensure that the cluster state on the primary relocation target has all
+            // replica shards that have recovered or are still recovering from the current primary, otherwise replication actions
+            // will not be send to these replicas. To accomplish this, first block new recoveries, then take version of latest cluster
+            // state. This means that no new recovery can be completed based on information of a newer cluster state than the current one.
+            newRecoveriesHandler.handle(new DelayRecoveryException("primary relocation hand-off in progress or completed for " + shardId));
```

Comment:
I think we should make newRecoveriesHanlder be:
```
 private final Function<String, Releasable> delayNewRecoveries; 
```

and then all of this becomes:

```
            try (Releasable ignored = delayNewRecoveries.apply("primary relocation hand-off in progress or completed for " + shardId)) {
                final long currentClusterStateVersion = currentClusterStateVersionSupplier.get();
                logger.trace("[{}][{}] waiting on {} to have cluster state with version [{}]", indexName, shardId, request.targetNode(),
                    currentClusterStateVersion);

                cancellableThreads.execute(() -> recoveryTarget.waitForClusterState(currentClusterStateVersion));

                logger.trace("[{}][{}] performing relocation hand-off to {}", indexName, shardId, request.targetNode());
                cancellableThreads.execute(() -> shard.relocated("to " + request.targetNode()));
            }
```